File: src/harness_core/error_taxonomy.py
"""Tool/Error Taxonomy schema and validation helpers.

Defines canonical error domains, the error_record schema, and validation
rules for classifying tool execution failures in the harness pipeline.
"""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from typing import Any, Dict, List, Optional, Sequence


# ---------------------------------------------------------------------------
# Canonical error domains
# ---------------------------------------------------------------------------

class ErrorDomain(str, Enum):
    """All recognised error domains in the taxonomy."""

    TOOL_CONTRACT_ERROR = "tool_contract_error"
    ENVIRONMENT_ERROR = "environment_error"
    CONTEXT_ERROR = "context_error"
    MODEL_JUDGMENT_ERROR = "model_judgment_error"
    EVALUATION_ERROR = "evaluation_error"
    HARNESS_BUG = "harness_bug"
    USER_ABORT = "user_abort"
    PROVIDER_ERROR = "provider_error"
    TIMEOUT = "timeout"
    UNKNOWN_ERROR = "unknown_error"


CANONICAL_DOMAINS: List[str] = [d.value for d in ErrorDomain]

# Domains that are never retryable
NON_RETRYABLE_DOMAINS: frozenset[str] = frozenset({
    ErrorDomain.USER_ABORT.value,
    ErrorDomain.HARNESS_BUG.value,
    ErrorDomain.UNKNOWN_ERROR.value,
})

# Domains where requires_human_triage is always True
MANDATORY_TRIAGE_DOMAINS: frozenset[str] = frozenset({
    ErrorDomain.UNKNOWN_ERROR.value,
    ErrorDomain.HARNESS_BUG.value,
})

# Domains that must never drive policy candidate adoption
NON_ADOPTABLE_DOMAINS: frozenset[str] = frozenset({
    ErrorDomain.UNKNOWN_ERROR.value,
})


# ---------------------------------------------------------------------------
# Schema version
# ---------------------------------------------------------------------------

SCHEMA_VERSION = "error_record.v1"

REQUIRED_FIELDS: Sequence[str] = (
    "schema_version",
    "error_id",
    "error_domain",
    "error_class",
    "retryable",
    "counts_against_model",
    "requires_human_triage",
    "tool_name",
    "model_profile_id",
    "context_pack_id",
    "event_id",
    "evidence_refs",
    "created_at",
)


# ---------------------------------------------------------------------------
# Data class
# ---------------------------------------------------------------------------

@dataclass
class ErrorRecord:
    """Typed representation of a single error_record."""

    error_id: str
    error_domain: str
    error_class: str
    retryable: bool
    counts_against_model: bool
    requires_human_triage: bool
    tool_name: str
    model_profile_id: str
    context_pack_id: str
    event_id: str
    evidence_refs: List[str]
    created_at: str
    schema_version: str = field(default=SCHEMA_VERSION)

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)

    def to_json(self, **kw: Any) -> str:
        return json.dumps(self.to_dict(), **kw)


# ---------------------------------------------------------------------------
# Validation helpers
# ---------------------------------------------------------------------------

def validate_error_record(data: Dict[str, Any]) -> List[str]:
    """Validate a dict against the error_record schema.

    Returns a list of violation strings.  Empty list means valid.
    """
    violations: List[str] = []

    # 1. Check all required fields are present
    for f in REQUIRED_FIELDS:
        if f not in data:
            violations.append(f"missing required field: {f}")

    if violations:
        return violations  # can't continue without required fields

    # 2. schema_version must match
    if data["schema_version"] != SCHEMA_VERSION:
        violations.append(
            f"schema_version must be {SCHEMA_VERSION}, got {data['schema_version']!r}"
        )

    # 3. error_domain must be canonical
    if data["error_domain"] not in CANONICAL_DOMAINS:
        violations.append(
            f"error_domain {data['error_domain']!r} not in canonical domains"
        )

    # 4. error_id must be a non-empty string
    if not isinstance(data["error_id"], str) or not data["error_id"]:
        violations.append("error_id must be a non-empty string")

    # 5. Boolean fields must be bool
    for bf in ("retryable", "counts_against_model", "requires_human_triage"):
        if not isinstance(data[bf], bool):
            violations.append(f"{bf} must be a bool, got {type(data[bf]).__name__}")

    # 6. evidence_refs must be a list
    if not isinstance(data["evidence_refs"], list):
        violations.append("evidence_refs must be a list")

    # 7. Domain-specific constraints
    domain = data["error_domain"]
    violations.extend(_validate_domain_constraints(data, domain))

    return violations


def _validate_domain_constraints(
    data: Dict[str, Any], domain: str
) -> List[str]:
    """Enforce domain-specific semantic rules."""
    violations: List[str] = []

    # unknown_error must always be non-retryable and require triage
    if domain == ErrorDomain.UNKNOWN_ERROR.value:
        if data["retryable"] is not False:
            violations.append(
                "unknown_error must have retryable=false (fail-hard)"
            )
        if data["requires_human_triage"] is not True:
            violations.append(
                "unknown_error must have requires_human_triage=true"
            )

    # user_abort is not a system failure — cannot be retryable
    if domain == ErrorDomain.USER_ABORT.value:
        if data["retryable"] is not False:
            violations.append("user_abort must have retryable=false")

    # harness_bug is not retryable and must require triage
    if domain == ErrorDomain.HARNESS_BUG.value:
        if data["retryable"] is not False:
            violations.append("harness_bug must have retryable=false")
        if data["requires_human_triage"] is not True:
            violations.append("harness_bug must have requires_human_triage=true")

    # provider_error and timeout must both be retryable by default
    if domain in (ErrorDomain.PROVIDER_ERROR.value, ErrorDomain.TIMEOUT.value):
        if data["retryable"] is not True:
            violations.append(
                f"{domain} should typically be retryable=true"
            )

    return violations


def create_error_record(
    *,
    error_domain: str,
    error_class: str,
    retryable: bool,
    counts_against_model: bool,
    requires_human_triage: bool,
    tool_name: str = "",
    model_profile_id: str = "",
    context_pack_id: str = "",
    event_id: str = "",
    evidence_refs: Optional[List[str]] = None,
    created_at: Optional[str] = None,
    error_id: Optional[str] = None,
) -> ErrorRecord:
    """Factory for ErrorRecord with auto-generated defaults."""
    return ErrorRecord(
        error_id=error_id or str(uuid.uuid4()),
        error_domain=error_domain,
        error_class=error_class,
        retryable=retryable,
        counts_against_model=counts_against_model,
        requires_human_triage=requires_human_triage,
        tool_name=tool_name,
        model_profile_id=model_profile_id,
        context_pack_id=context_pack_id,
        event_id=event_id,
        evidence_refs=evidence_refs or [],
        created_at=created_at or datetime.now(timezone.utc).isoformat(),
    )


def load_fixture(path: Path) -> Dict[str, Any]:
    """Load a single JSON fixture file and return the dict."""
    with open(path, "r") as f:
        return json.load(f)


def load_and_validate_fixture(path: Path) -> tuple[Dict[str, Any], List[str]]:
    """Load a fixture and return (data, violations)."""
    data = load_fixture(path)
    return data, validate_error_record(data)


def load_all_fixtures(fixture_dir: Path) -> List[tuple[str, Dict[str, Any], List[str]]]:
    """Load every .json fixture in a directory.

    Returns list of (filename, data, violations) tuples.
    """
    results = []
    for p in sorted(fixture_dir.glob("*.json")):
        data = load_fixture(p)
        violations = validate_error_record(data)
        results.append((p.name, data, violations))
    return results


def is_adoptable(domain: str) -> bool:
    """Return True if errors in this domain may drive policy candidate adoption."""
    return domain not in NON_ADOPTABLE_DOMAINS


def is_retryable(domain: str) -> bool:
    """Return True if errors in this domain are expected to be retryable."""
    return domain not in NON_RETRYABLE_DOMAINS


def requires_triage(domain: str) -> bool:
    """Return True if errors in this domain always require human triage."""
    return domain in MANDATORY_TRIAGE_DOMAINS

File: engine/src/app_layer/error_taxonomy.rs
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

pub const SCHEMA_VERSION: &str = "error_record.v1";

pub const CANONICAL_DOMAINS: &[&str] = &[
    "tool_contract_error",
    "environment_error",
    "context_error",
    "model_judgment_error",
    "evaluation_error",
    "harness_bug",
    "user_abort",
    "provider_error",
    "timeout",
    "unknown_error",
];

pub const REQUIRED_FIELDS: &[&str] = &[
    "schema_version",
    "error_id",
    "error_domain",
    "error_class",
    "retryable",
    "counts_against_model",
    "requires_human_triage",
    "tool_name",
    "model_profile_id",
    "context_pack_id",
    "event_id",
    "evidence_refs",
    "created_at",
];

pub const NON_RETRYABLE_DOMAINS: &[&str] = &["user_abort", "harness_bug", "unknown_error"];
pub const MANDATORY_TRIAGE_DOMAINS: &[&str] = &["unknown_error", "harness_bug"];
pub const NON_ADOPTABLE_DOMAINS: &[&str] = &["unknown_error"];

// ---------------------------------------------------------------------------
// ErrorDomain
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ErrorDomain {
    ToolContractError,
    EnvironmentError,
    ContextError,
    ModelJudgmentError,
    EvaluationError,
    HarnessBug,
    UserAbort,
    ProviderError,
    Timeout,
    UnknownError,
}

impl ErrorDomain {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::ToolContractError => "tool_contract_error",
            Self::EnvironmentError => "environment_error",
            Self::ContextError => "context_error",
            Self::ModelJudgmentError => "model_judgment_error",
            Self::EvaluationError => "evaluation_error",
            Self::HarnessBug => "harness_bug",
            Self::UserAbort => "user_abort",
            Self::ProviderError => "provider_error",
            Self::Timeout => "timeout",
            Self::UnknownError => "unknown_error",
        }
    }

    pub fn parse_domain(s: &str) -> Option<Self> {
        match s {
            "tool_contract_error" => Some(Self::ToolContractError),
            "environment_error" => Some(Self::EnvironmentError),
            "context_error" => Some(Self::ContextError),
            "model_judgment_error" => Some(Self::ModelJudgmentError),
            "evaluation_error" => Some(Self::EvaluationError),
            "harness_bug" => Some(Self::HarnessBug),
            "user_abort" => Some(Self::UserAbort),
            "provider_error" => Some(Self::ProviderError),
            "timeout" => Some(Self::Timeout),
            "unknown_error" => Some(Self::UnknownError),
            _ => None,
        }
    }
}

// ---------------------------------------------------------------------------
// ErrorRecord
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct ErrorRecord {
    pub error_id: String,
    pub error_domain: String,
    pub error_class: String,
    pub retryable: bool,
    pub counts_against_model: bool,
    pub requires_human_triage: bool,
    pub tool_name: String,
    pub model_profile_id: String,
    pub context_pack_id: String,
    pub event_id: String,
    pub evidence_refs: Vec<String>,
    pub created_at: String,
    pub schema_version: String,
}

impl Default for ErrorRecord {
    fn default() -> Self {
        Self {
            error_id: String::new(),
            error_domain: String::new(),
            error_class: String::new(),
            retryable: false,
            counts_against_model: false,
            requires_human_triage: false,
            tool_name: String::new(),
            model_profile_id: String::new(),
            context_pack_id: String::new(),
            event_id: String::new(),
            evidence_refs: Vec::new(),
            created_at: String::new(),
            schema_version: SCHEMA_VERSION.to_string(),
        }
    }
}

impl ErrorRecord {
    pub fn to_value(&self) -> serde_json::Value {
        serde_json::to_value(self).unwrap()
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

pub fn validate_error_record(data: &serde_json::Value) -> Vec<String> {
    let mut violations = Vec::new();

    let object = match data.as_object() {
        Some(o) => o,
        None => {
            violations.push("error_record must be a JSON object".to_string());
            return violations;
        }
    };

    for field in REQUIRED_FIELDS {
        if !object.contains_key(*field) {
            violations.push(format!("missing required field: {}", field));
        }
    }
    if !violations.is_empty() {
        return violations;
    }

    let sv = object["schema_version"].as_str().unwrap_or("");
    if sv != SCHEMA_VERSION {
        violations.push(format!(
            "schema_version must be {}, got {:?}",
            SCHEMA_VERSION, sv
        ));
    }

    let domain = object["error_domain"].as_str().unwrap_or("");
    if !CANONICAL_DOMAINS.contains(&domain) {
        violations.push(format!(
            "error_domain {:?} not in canonical domains",
            domain
        ));
    }

    match object.get("error_id").and_then(serde_json::Value::as_str) {
        Some(s) if !s.is_empty() => {}
        _ => violations.push("error_id must be a non-empty string".to_string()),
    }

    for bf in &["retryable", "counts_against_model", "requires_human_triage"] {
        if !object[*bf].is_boolean() {
            violations.push(format!(
                "{} must be a bool, got {}",
                bf,
                json_type_name(&object[*bf])
            ));
        }
    }

    if !object["evidence_refs"].is_array() {
        violations.push("evidence_refs must be a list".to_string());
    }

    violations.extend(validate_domain_constraints(object, domain));

    violations
}

fn validate_domain_constraints(
    object: &serde_json::Map<String, serde_json::Value>,
    domain: &str,
) -> Vec<String> {
    let mut violations = Vec::new();

    let retryable = object.get("retryable").and_then(serde_json::Value::as_bool);
    let triage = object
        .get("requires_human_triage")
        .and_then(serde_json::Value::as_bool);

    if domain == "unknown_error" {
        if retryable != Some(false) {
            violations.push("unknown_error must have retryable=false (fail-hard)".to_string());
        }
        if triage != Some(true) {
            violations.push("unknown_error must have requires_human_triage=true".to_string());
        }
    }

    if domain == "user_abort" && retryable != Some(false) {
        violations.push("user_abort must have retryable=false".to_string());
    }

    if domain == "harness_bug" {
        if retryable != Some(false) {
            violations.push("harness_bug must have retryable=false".to_string());
        }
        if triage != Some(true) {
            violations.push("harness_bug must have requires_human_triage=true".to_string());
        }
    }

    if (domain == "provider_error" || domain == "timeout") && retryable != Some(true) {
        violations.push(format!("{} should typically be retryable=true", domain));
    }

    violations
}

fn json_type_name(v: &serde_json::Value) -> &'static str {
    match v {
        serde_json::Value::Null => "null",
        serde_json::Value::Bool(_) => "bool",
        serde_json::Value::Number(_) => "number",
        serde_json::Value::String(_) => "string",
        serde_json::Value::Array(_) => "array",
        serde_json::Value::Object(_) => "object",
    }
}

// ---------------------------------------------------------------------------
// Factory
// ---------------------------------------------------------------------------

pub struct CreateErrorRecordParams<'a> {
    pub error_domain: &'a str,
    pub error_class: &'a str,
    pub retryable: bool,
    pub counts_against_model: bool,
    pub requires_human_triage: bool,
    pub tool_name: &'a str,
    pub model_profile_id: &'a str,
    pub context_pack_id: &'a str,
    pub event_id: &'a str,
    pub evidence_refs: Option<Vec<String>>,
    pub created_at: Option<&'a str>,
    pub error_id: Option<&'a str>,
}

pub fn create_error_record(params: CreateErrorRecordParams<'_>) -> ErrorRecord {
    ErrorRecord {
        error_id: params
            .error_id
            .map(str::to_string)
            .unwrap_or_else(|| uuid::Uuid::new_v4().to_string()),
        error_domain: params.error_domain.to_string(),
        error_class: params.error_class.to_string(),
        retryable: params.retryable,
        counts_against_model: params.counts_against_model,
        requires_human_triage: params.requires_human_triage,
        tool_name: params.tool_name.to_string(),
        model_profile_id: params.model_profile_id.to_string(),
        context_pack_id: params.context_pack_id.to_string(),
        event_id: params.event_id.to_string(),
        evidence_refs: params.evidence_refs.unwrap_or_default(),
        created_at: params
            .created_at
            .map(str::to_string)
            .unwrap_or_else(|| chrono::Utc::now().to_rfc3339()),
        schema_version: SCHEMA_VERSION.to_string(),
    }
}

// ---------------------------------------------------------------------------
// Classification helpers
// ---------------------------------------------------------------------------

pub fn is_adoptable(domain: &str) -> bool {
    !NON_ADOPTABLE_DOMAINS.contains(&domain)
}

pub fn is_retryable(domain: &str) -> bool {
    !NON_RETRYABLE_DOMAINS.contains(&domain)
}

pub fn requires_triage(domain: &str) -> bool {
    MANDATORY_TRIAGE_DOMAINS.contains(&domain)
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn valid_record_json() -> serde_json::Value {
        json!({
            "schema_version": SCHEMA_VERSION,
            "error_id": "err-001",
            "error_domain": "provider_error",
            "error_class": "rate_limit",
            "retryable": true,
            "counts_against_model": true,
            "requires_human_triage": false,
            "tool_name": "code_gen",
            "model_profile_id": "mimo-v2",
            "context_pack_id": "cp-001",
            "event_id": "evt-001",
            "evidence_refs": ["ref-1"],
            "created_at": "2026-01-01T00:00:00Z"
        })
    }

    #[test]
    fn test_valid_record_passes() {
        let violations = validate_error_record(&valid_record_json());
        assert!(
            violations.is_empty(),
            "expected no violations, got {:?}",
            violations
        );
    }

    #[test]
    fn test_missing_required_field() {
        let mut data = valid_record_json();
        data.as_object_mut().unwrap().remove("error_id");
        let violations = validate_error_record(&data);
        assert!(violations.iter().any(|v| v.contains("error_id")));
    }

    #[test]
    fn test_bad_schema_version() {
        let mut data = valid_record_json();
        data["schema_version"] = json!("wrong.v1");
        let violations = validate_error_record(&data);
        assert!(violations.iter().any(|v| v.contains("schema_version")));
    }

    #[test]
    fn test_unknown_error_must_be_non_retryable_and_triage() {
        let mut data = valid_record_json();
        data["error_domain"] = json!("unknown_error");
        data["retryable"] = json!(true);
        data["requires_human_triage"] = json!(false);
        let violations = validate_error_record(&data);
        assert!(violations.iter().any(|v| v.contains("retryable=false")));
        assert!(violations
            .iter()
            .any(|v| v.contains("requires_human_triage=true")));
    }

    #[test]
    fn test_harness_bug_constraints() {
        let mut data = valid_record_json();
        data["error_domain"] = json!("harness_bug");
        data["retryable"] = json!(true);
        data["requires_human_triage"] = json!(false);
        let violations = validate_error_record(&data);
        assert!(violations
            .iter()
            .any(|v| v.contains("harness_bug must have retryable=false")));
        assert!(violations
            .iter()
            .any(|v| v.contains("harness_bug must have requires_human_triage=true")));
    }

    #[test]
    fn test_user_abort_not_retryable() {
        let mut data = valid_record_json();
        data["error_domain"] = json!("user_abort");
        data["retryable"] = json!(true);
        let violations = validate_error_record(&data);
        assert!(violations
            .iter()
            .any(|v| v.contains("user_abort must have retryable=false")));
    }

    #[test]
    fn test_provider_error_should_be_retryable() {
        let mut data = valid_record_json();
        data["error_domain"] = json!("provider_error");
        data["retryable"] = json!(false);
        let violations = validate_error_record(&data);
        assert!(violations
            .iter()
            .any(|v| v.contains("should typically be retryable=true")));
    }

    #[test]
    fn test_domain_classification_helpers() {
        assert!(is_adoptable("provider_error"));
        assert!(!is_adoptable("unknown_error"));

        assert!(is_retryable("provider_error"));
        assert!(!is_retryable("unknown_error"));
        assert!(!is_retryable("user_abort"));
        assert!(!is_retryable("harness_bug"));

        assert!(requires_triage("unknown_error"));
        assert!(requires_triage("harness_bug"));
        assert!(!requires_triage("provider_error"));
    }

    #[test]
    fn test_create_error_record_defaults() {
        let rec = create_error_record(CreateErrorRecordParams {
            error_domain: "timeout",
            error_class: "deadline_exceeded",
            retryable: true,
            counts_against_model: false,
            requires_human_triage: false,
            tool_name: "tool_x",
            model_profile_id: "profile_y",
            context_pack_id: "ctx_z",
            event_id: "evt-1",
            evidence_refs: None,
            created_at: Some("2026-01-01T00:00:00Z"),
            error_id: Some("err-fixed"),
        });
        assert_eq!(rec.error_id, "err-fixed");
        assert_eq!(rec.error_domain, "timeout");
        assert_eq!(rec.schema_version, SCHEMA_VERSION);
        assert!(rec.evidence_refs.is_empty());
    }

    #[test]
    fn test_error_domain_roundtrip() {
        for domain_str in CANONICAL_DOMAINS {
            let domain = ErrorDomain::parse_domain(domain_str).unwrap();
            assert_eq!(domain.as_str(), *domain_str);
        }
    }

    #[test]
    fn test_error_record_to_value() {
        let rec = create_error_record(CreateErrorRecordParams {
            error_domain: "environment_error",
            error_class: "missing_dep",
            retryable: false,
            counts_against_model: true,
            requires_human_triage: false,
            tool_name: "t",
            model_profile_id: "p",
            context_pack_id: "c",
            event_id: "e",
            evidence_refs: Some(vec!["r1".to_string()]),
            created_at: Some("2026-01-01T00:00:00Z"),
            error_id: Some("err-001"),
        });
        let v = rec.to_value();
        assert_eq!(v["error_domain"], "environment_error");
        assert_eq!(v["schema_version"], SCHEMA_VERSION);
        assert_eq!(v["evidence_refs"][0], "r1");
    }
}
